---
title: "Homework 3"
author: "Yavuz Delen"
date: "4/14/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```


```{r}
loblolly_pheno <- read.csv("data/DATA_rootnum_age10_rootnum.csv", header = TRUE, stringsAsFactors = FALSE)
#View(loblolly_pheno)


geno_data <- "https://jyanglab.com/img/data/Snp_Data.csv"

loblolly_geno <- read.csv(geno_data, header=TRUE, stringsAsFactors = FALSE)
head(loblolly_geno)
#View(loblolly_geno)

dim(loblolly_pheno) #926 4
dim(loblolly_geno) #926 4854
loblolly_geno[1:5, 1:5]
```


## Remove missing phenotypes
```{r, eval=FALSE}
is.na(loblolly_pheno$Derregressed_BV)
#sum(is.na(loblolly_pheno$Derregressed_BV)) #1
na.index <-  which(is.na(loblolly_pheno$Derregressed_BV))
#head(loblolly_pheno[na.index,])


# length(na.index) #1
loblolly_pheno <- loblolly_pheno[-na.index, ]


# Keep genotypes for these remaining lines
loblolly_geno <- loblolly_geno[loblolly_geno$Genotype %in% loblolly_pheno$Genotype, ]

# phenotypes 
y <- loblolly_pheno$Derregressed_BV
y <- matrix(y, ncol=1)
y

# markers 
loblolly_geno <- loblolly_geno[,-1] 
loblolly_geno[loblolly_geno == -9] <- NA

```


# SNP quality control
### Missingness and MAF

```{r, eval=FALSE, echo=TRUE}
# missing rate
missing <- apply(loblolly_geno, 2, function(x){sum(is.na(x))/length(x)})
head(missing)
#hist(missing)


# minor allele frequency
maf <- apply(loblolly_geno, 2, function(x){
  frq <- mean(x, na.rm=TRUE)/2
  return(ifelse(frq > 0.5, 1-frq, frq))
})
```

#### Plot the results
```{r, eval=FALSE, echo=TRUE}
hist(missing, breaks=100, col="blue", xlab="SNP Missing rate")
hist(maf, breaks=100, col="blue", xlab="Minor Allele Freq")
```


# SNP quality control

Removing SNPs with high missing rate (missingness > 0.2) and low MAF (MAF < 0.05)

```{r, eval=FALSE, echo=TRUE}
idx1 <- which(missing > 0.2) #154
idx2 <- which(maf < 0.05) #1647
idx <- unique(c(idx1, idx2)) #1784

loblolly_geno2 <- loblolly_geno[, -idx]
dim(loblolly_geno2)  #925 3074
```


##burada kaldim
### Missing marker imputation

Replace missing marker genotypes with __mean values__. Then store the marker genotypes in a matrix object `Z`. 

```{r, eval=FALSE, echo=TRUE}
Z <- matrix(0, ncol=ncol(loblolly_geno2), nrow=nrow(loblolly_geno2))
for (j in 1:ncol(loblolly_geno2)){
  #cat("j = ", j, '\n')
  Z[,j] <- ifelse(is.na(loblolly_geno2[,j]), mean(loblolly_geno2[,j], na.rm=TRUE), loblolly_geno2[,j])
}
# sum(is.na(Z))
```


# Genomic relationship

### SNP Matrix standardization

Standardize the genotype matrix to have a mean of zero and variance of one. Save this matrix as `Zs`. 

```{r, eval=FALSE, echo=TRUE}
Zs <- scale(Z, center = TRUE, scale = TRUE)
# dimensions 
n <- nrow(Zs)
m <- ncol(Zs)
```


### Calcualte genomic relationship
 

```{r, eval=FALSE, echo=TRUE}
# Given matrices x and y as arguments, return a matrix cross-product. This is formally equivalent to (but usually slightly faster than) the call t(x) %*% y (crossprod) or x %*% t(y) (tcrossprod).
G <- tcrossprod(Zs) / ncol(Zs)
G <- G + diag(n)*0.001
```



# Solve MME for GBLUP

```{r, eval=FALSE, echo=TRUE}
lambda <- 1.35 # fit$Ve / fit$Vm
Ginv <- solve(G)
ones <- matrix(1, ncol=1, nrow=n)
Z <- diag(n)
# Given matrices x and y as arguments, return a matrix cross-product. This is formally equivalent to (but usually slightly faster than) the call t(x) %*% y (crossprod) or x %*% t(y) (tcrossprod).
LHS1 <- cbind(crossprod(ones), crossprod(ones, Z)) 
LHS2 <- cbind(crossprod(Z, ones), crossprod(Z) +  Ginv*lambda)
LHS <- rbind(LHS1, LHS2)
RHS <- rbind( crossprod(ones, y), crossprod(Z,y) )
sol <- solve(LHS, RHS)
head(sol)
tail(sol)
```



```{r, eval=FALSE, echo=TRUE}
install.packages("rrBLUP")
library(rrBLUP)
fit <- mixed.solve(y = y, K=G)
# additive genetic variance
fit$Vu
# residual variance
fit$Ve
# intercept 
fit$beta
# additive genetic values
head(fit$u)
tail(fit$u)
# genomic h2
fit$Vu / (fit$Vu + fit$Ve)
# ratio of variance components 
fit$Ve / fit$Vu
plot(x=sol[-1], y=fit$u)
```



```{r, eval=FALSE, echo=TRUE}
lambda <- 4326.212 # fit$Ve / fit$Vu
ones <- matrix(1, ncol=1, nrow=n)
I <- diag(m)
LHS1 <- cbind(crossprod(ones), crossprod(ones, Zs)) 
LHS2 <- cbind(crossprod(Zs, ones), crossprod(Zs) +  I*lambda)
LHS <- rbind(LHS1, LHS2)
RHS <- rbind( crossprod(ones, y), crossprod(Zs,y) )
sol2 <- solve(LHS, RHS)
head(sol2)
tail(sol2)
```


```{r, eval=FALSE, echo=TRUE}
library(rrBLUP)
fit2 <- mixed.solve(y = y, Z=Zs)
# marker additive genetic variance
fit2$Vu
# residual variance
fit2$Ve
# intercept 
fit2$beta
# marker additive genetic effects
head(fit2$u)
tail(fit2$u)
# ratio of variance components 
fit2$Ve / fit2$Vu

plot(x=sol2[-1], y=fit2$u)
```
